geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala

```diff
@@ -0,0 +1,122 @@
+class GeoAvroRecordSetWriterFactory extends AbstractControllerService with RecordSetWriterFactory {
+  // NB: This is the same as what the InheritSchemaFromRecord Strategy does
+  override def getSchema(map: util.Map[String, String], recordSchema: RecordSchema): RecordSchema = recordSchema
+
+  override def createWriter(componentLog: ComponentLog,
+                            recordSchema: RecordSchema,
+                            outputStream: OutputStream,
+                            map: util.Map[String, String]): GeoAvroRecordSetWriter = {
+    new GeoAvroRecordSetWriter(componentLog, recordSchema, outputStream, getConfigurationContext.getProperties)
+  }
+
+  override def getSupportedPropertyDescriptors: util.List[PropertyDescriptor] =
+    Seq(GEOMETRY_COLUMNS, IncludeVisibilities).toList.asJava
+}
+
+object GeoAvroRecordSetWriterFactory {
+  // TODO factor out
+  val IncludeVisibilities: PropertyDescriptor =
+    new PropertyDescriptor.Builder()
+      .name("include-visibilities")
+      .displayName("Include Visibilities")
+      .description("Include a column with visibility expressions for each row")
+      .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+      .allowableValues("false", "true")
+      .defaultValue("false")
+      .build
+
+  val GEOMETRY_COLUMNS: PropertyDescriptor = new PropertyDescriptor.Builder()
+    .name("Geometry Columns")
+    .description("Comma-separated list of columns with geometries with type. " +
+      "Example: position:Point,line:LineString.  " +
+      "Note the first field is used as a the default geometry.")
+    .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .required(false).build
+
+  val TYPE_NAME: PropertyDescriptor = new PropertyDescriptor.Builder()
+    .name("SimpleFeature TypeName")
+    .description("The type name for the output.")
+    .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .required(false).build
+}
+
+class GeoAvroRecordSetWriter(componentLog: ComponentLog, recordSchema: RecordSchema, outputStream: OutputStream, map: util.Map[PropertyDescriptor, String]) extends AbstractRecordSetWriter(outputStream) {
+  private val schemaAccessWriter = new SchemaNameAsAttribute()
+  // Wkt is assumed to be the default GeometryEncoding.
+  // TODO:  Extract this as a property?
+  private val encodings = getEncodings(map, GeometryEncoding.Wkt)
+  private val defaultGeometryColumn = Option(map.get(GEOMETRY_COLUMNS)).map(_.split(":")(0))
+  private val typeName = Option(map.get(TYPE_NAME))
+
+  private val geometryColumns= encodings.map { case (k, v) =>
+    GeometryColumn(k, v.clazz, defaultGeometryColumn.isDefined && defaultGeometryColumn.get.equals(k))
+  }.toSeq
+
+  // TODO Abstract
+  private val visField = if (java.lang.Boolean.parseBoolean(map.get(IncludeVisibilities))) {
+      Some("visibilities")
```

Comment:
@elahrvivaz  Q:  How do we want to handle the visibilities field in GetGeoMesaKafkaRecord and the GeoAvroRecordSetWriterFactory?  The former has a boolean that when true leads to a field named "visibilities" being used.  Should the GeoAvroWriter do the same?  Or allow for a configurable field?

Reply:
my thought was on conversion to a record (i.e. in GetGeoMesaKafkaRecord) we can go with standardized names. On conversion from a record (i.e. in PutGeoMesa*Record) we should support a configurable name. (one can always use nifi record processors to rename and manipulate the record cols)

Reply:
also see https://github.com/geomesa/geomesa-nifi/pull/87 for how i handled the user data. note there will be overlap if you include user data and vis, i didn't think it mattered enough to worry about it though.